src/v/cloud_storage/remote_segment.cc

```diff
@@ -886,12 +913,26 @@ ss::future<> remote_segment::hydrate() {
 
     auto g = _gate.hold();
     vlog(_ctxlog.debug, "segment {} hydration requested", _path);
-    ss::promise<ss::file> p;
-    auto fut = p.get_future();
+    ss::shared_promise<ss::file> p;
```

Comment:
i'm not seeing why we need to have a shared promise / future. sometimes it's necessary, but it has been a source of problems in the past and it can often be avoided. why is it needed here?

Reply:
Shared promise gives us two methods that we use here, overloads of get_future, one that ties a future to an abort source and one that ties a future to a timeout. We use both to abort the wait for hydration if client disconnects or a timeout occurs.

Since we're not using the `available` and `failed` query methods of shared promise which can be unreliable (and seems to be fixed very recently in seastar), I think we should be safe. Are there other issues with shared promise? This is the one I know of so this PR doesn't use those methods.